core/templates/components/question-directives/question-player/services/question-player-state.service.ts

```diff
@@ -62,53 +62,71 @@ export class QuestionPlayerStateService {
       linkedSkillIds: linkedSkillIds,
       answers: [],
       usedHints: [],
-      viewedSolution: null
+      viewedSolution: undefined
     };
   }
 
   hintUsed(question: Question): void {
     let questionId = question.getId();
-    if (!this.questionPlayerState[questionId]) {
-      this._createNewQuestionPlayerState(
-        questionId, question.getLinkedSkillIds());
+    // QuestionId null signifies temporary question and it does
+    // not have associated question player state.
+    if (questionId !== null) {
+      // Typecasting is done just to deal with TS error.
+      questionId = String(questionId);
```

Comment:
Why is this needed when above we check if it is not `null`.

Reply:
ditto below

Reply:
@vojtechjelinek questionID stores value of the id of the question which is defined as string | null return type from getID function, as null specifies a temporary question the null check is there so that it's state related properties are not set for a temporary question. Although this scenerio should never arise.

Reply:
Yeah, I know but that doesn't explain why `String()` is needed here. AFAIK TS compiler should be smart enough to know that `questionId` cannot be `null` after the if.

Reply:
@vojtechjelinek When we apply strict TS type checks, how it currently works is that a variable of type `string|null` cannot be assigned to a variable of type `string`, it does not matter if we have a conditional handling for that null case or not. That's why it has to be there or else it will give a strict type check error. I saw it has been implemented other way like
`let questionId = question.getId() as string;` to resolve such issues. Will make it same as them. Thanks!

Reply:
@vojtechjelinek Made this change.